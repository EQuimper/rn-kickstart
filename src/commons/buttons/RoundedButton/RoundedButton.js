// @flow

import * as React from 'react';
import { StyleSheet, ActivityIndicator } from 'react-native';
import Touchable from '@appandflow/touchable';
import { type StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';
import Color from 'color';
import invariant from 'invariant';

import { theme } from '../../../utils';
import { Callout } from '../../Typographie';

type P = {|
  style?: StyleObj,
  textStyle?: StyleObj,
  onPress: () => void,
  children: React.Node,
  disabled?: boolean,
  loading?: boolean,
  variant?: 'primary' | 'secondary',
  color?: string,
  loadingColor?: string,
|};

function RoundedButton({
  style,
  onPress,
  children,
  disabled = false,
  loading = false,
  variant,
  color,
  loadingColor,
  textStyle,
  ...rest
}: P) {
  const colorNotProvided = !color && !variant;

  invariant(
    !colorNotProvided,
    'RoundedButton: You must provided at least one of variant or color',
  );

  const isDisabled = disabled || loading;

  const _color = Color(variant ? theme.colors[variant] : color);

  const backgroundColor = isDisabled ? _color.lighten(0.5) : _color;

  return (
    <Touchable
      {...rest}
      style={[styles.root, { backgroundColor }, style]}
      onPress={onPress}
      feedback="opacity"
      disabled={isDisabled}
    >
      {loading ? (
        <ActivityIndicator color={loadingColor} size="small" />
      ) : (
        <Callout style={[styles.text, textStyle]}>{children}</Callout>
      )}
    </Touchable>
  );
}

const styles = StyleSheet.create({
  root: {
    height: 40,
    width: '100%',
    borderRadius: theme.metrics.radius.smallSize,
    backgroundColor: 'red',
    alignItems: 'center',
    justifyContent: 'center',
    ...theme.metrics.shadow.basic,
  },
  text: {
    color: theme.colors.white,
  },
});

export default RoundedButton;
