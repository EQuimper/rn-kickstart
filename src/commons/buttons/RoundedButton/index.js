// @flow

export { default as RoundedButton } from './RoundedButton';
