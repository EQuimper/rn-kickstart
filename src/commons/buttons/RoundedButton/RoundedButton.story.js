// @flow

import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { action } from '@storybook/addon-actions';

import { CenterView } from '../../';
import RoundedButton from './RoundedButton';

storiesOf('RoundedButton', module)
  .addDecorator(getStory => (
    <CenterView style={{ paddingHorizontal: 20 }}>{getStory()}</CenterView>
  ))
  .add('Default', () => (
    <RoundedButton variant="primary" onPress={action('clicked-text')}>
      Hello World
    </RoundedButton>
  ))
  .add('Disabled', () => (
    <RoundedButton disabled variant="primary" onPress={action('clicked-text')}>
      Hello World
    </RoundedButton>
  ))
  .add('Loading', () => (
    <RoundedButton
      loadingColor="#fff"
      loading
      variant="primary"
      onPress={action('clicked-text')}
    >
      Hello World
    </RoundedButton>
  ));
