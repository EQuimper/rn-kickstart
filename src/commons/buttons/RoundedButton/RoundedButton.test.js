// @noflow

import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { ActivityIndicator } from 'react-native';

import RoundedButton from './RoundedButton';

describe('<RoundedButton />', () => {
  it('should renders correctly', () => {
    const props = {
      onPress: () => null,
      variant: 'primary',
    };

    const tree = renderer
      .create(<RoundedButton {...props}>Hello World</RoundedButton>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should handle press event', () => {
    const spy = jest.fn();
    const text = 'Hello World';

    const props = {
      onPress: spy,
      variant: 'primary',
    };

    const button = shallow(<RoundedButton {...props}>{text}</RoundedButton>);

    button.simulate('press');
    expect(spy).toHaveBeenCalled();
  });

  it('should contain the text pass as children', () => {
    const text = 'Hello World';

    const props = {
      onPress: () => null,
      variant: 'primary',
    };

    const button = shallow(<RoundedButton {...props}>{text}</RoundedButton>);
    expect(button.contains(text)).toBeTruthy();
  });

  it('should be able to change textStyle', () => {
    const text = 'Hello World';

    const props = {
      onPress: () => null,
      variant: 'primary',
      textStyle: {
        color: 'red',
        fontSize: 20,
      },
    };

    const button = renderer
      .create(<RoundedButton {...props}>{text}</RoundedButton>)
      .toJSON();
    expect(button).toMatchSnapshot();
  });

  it('should be in disabled mode if disabled true', () => {
    const text = 'Hello World';

    const props = {
      onPress: () => null,
      variant: 'primary',
    };

    const buttonDisabled = renderer
      .create(
        <RoundedButton {...props} disabled>
          {text}
        </RoundedButton>,
      )
      .toJSON();
    expect(buttonDisabled).toMatchSnapshot();
  });

  it('should be in disabled mode if loading true and show ActivityIndicator', () => {
    const text = 'Hello World';

    const props = {
      onPress: () => null,
      variant: 'primary',
    };

    const buttonLoading = shallow(
      <RoundedButton {...props} loading>
        {text}
      </RoundedButton>,
    );

    expect(buttonLoading.find(ActivityIndicator)).toHaveLength(1);
    expect(buttonLoading).toMatchSnapshot();
  });

  it('should throw error if variant or color not passed as props', () => {
    const text = 'Hello World';

    const props = {
      onPress: () => null,
    };

    const buttonWithError = () =>
      shallow(<RoundedButton {...props}>{text}</RoundedButton>);

    expect(buttonWithError).toThrow(
      'RoundedButton: You must provided at least one of variant or color',
    );
    expect(buttonWithError).toThrowErrorMatchingSnapshot();
  });
});
