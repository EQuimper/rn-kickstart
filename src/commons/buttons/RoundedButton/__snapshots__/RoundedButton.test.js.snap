// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoundedButton /> should be able to change textStyle 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={false}
  feedback="opacity"
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "red",
        "borderRadius": 16,
        "elevation": 2,
        "height": 40,
        "justifyContent": "center",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 13,
          "width": 0,
        },
        "shadowOpacity": 0.12,
        "shadowRadius": 25,
        "width": "100%",
      },
      Object {
        "backgroundColor": Object {
          "color": Array [
            233,
            82,
            128,
          ],
          "model": "rgb",
          "valpha": 1,
        },
      },
      undefined,
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#000",
          "fontSize": 16,
          "fontWeight": "400",
        },
        Array [
          Object {
            "color": "#fff",
          },
          Object {
            "color": "red",
            "fontSize": 20,
          },
        ],
      ]
    }
  >
    Hello World
  </Text>
</TouchableOpacity>
`;

exports[`<RoundedButton /> should be in disabled mode if disabled true 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={true}
  feedback="opacity"
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "red",
        "borderRadius": 16,
        "elevation": 2,
        "height": 40,
        "justifyContent": "center",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 13,
          "width": 0,
        },
        "shadowOpacity": 0.12,
        "shadowRadius": 25,
        "width": "100%",
      },
      Object {
        "backgroundColor": Object {
          "color": Array [
            341.7218543046358,
            77.4358974358974,
            92.64705882352942,
          ],
          "model": "hsl",
          "valpha": 1,
        },
      },
      undefined,
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#000",
          "fontSize": 16,
          "fontWeight": "400",
        },
        Array [
          Object {
            "color": "#fff",
          },
          undefined,
        ],
      ]
    }
  >
    Hello World
  </Text>
</TouchableOpacity>
`;

exports[`<RoundedButton /> should be in disabled mode if loading true and show ActivityIndicator 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RoundedButton
    loading={true}
    onPress={[Function]}
    variant="primary"
  >
    Hello World
  </RoundedButton>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activeOpacity": 0.2,
      "children": <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="small"
      />,
      "disabled": true,
      "feedback": "opacity",
      "onPress": [Function],
      "style": Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "red",
          "borderRadius": 16,
          "elevation": 2,
          "height": 40,
          "justifyContent": "center",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 13,
            "width": 0,
          },
          "shadowOpacity": 0.12,
          "shadowRadius": 25,
          "width": "100%",
        },
        Object {
          "backgroundColor": Object {
            "color": Array [
              341.7218543046358,
              77.4358974358974,
              92.64705882352942,
            ],
            "model": "hsl",
            "valpha": 1,
          },
        },
        undefined,
      ],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "animating": true,
        "color": "#999999",
        "hidesWhenStopped": true,
        "size": "small",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeOpacity": 0.2,
        "children": <ActivityIndicator
          animating={true}
          color="#999999"
          hidesWhenStopped={true}
          size="small"
        />,
        "disabled": true,
        "feedback": "opacity",
        "onPress": [Function],
        "style": Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "red",
            "borderRadius": 16,
            "elevation": 2,
            "height": 40,
            "justifyContent": "center",
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 13,
              "width": 0,
            },
            "shadowOpacity": 0.12,
            "shadowRadius": 25,
            "width": "100%",
          },
          Object {
            "backgroundColor": Object {
              "color": Array [
                341.7218543046358,
                77.4358974358974,
                92.64705882352942,
              ],
              "model": "hsl",
              "valpha": 1,
            },
          },
          undefined,
        ],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animating": true,
          "color": "#999999",
          "hidesWhenStopped": true,
          "size": "small",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<RoundedButton /> should renders correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={false}
  feedback="opacity"
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "red",
        "borderRadius": 16,
        "elevation": 2,
        "height": 40,
        "justifyContent": "center",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 13,
          "width": 0,
        },
        "shadowOpacity": 0.12,
        "shadowRadius": 25,
        "width": "100%",
      },
      Object {
        "backgroundColor": Object {
          "color": Array [
            233,
            82,
            128,
          ],
          "model": "rgb",
          "valpha": 1,
        },
      },
      undefined,
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#000",
          "fontSize": 16,
          "fontWeight": "400",
        },
        Array [
          Object {
            "color": "#fff",
          },
          undefined,
        ],
      ]
    }
  >
    Hello World
  </Text>
</TouchableOpacity>
`;

exports[`<RoundedButton /> should throw error if variant or color not passed as props 1`] = `"RoundedButton: You must provided at least one of variant or color"`;
